File: backend/src/main/java/com/joelcode/personalinvestmentportfoliotracker/services/order/OrderService.java
package com.joelcode.personalinvestmentportfoliotracker.services.order;

import com.joelcode.personalinvestmentportfoliotracker.dto.order.CreateOrderRequest;
import com.joelcode.personalinvestmentportfoliotracker.dto.order.OrderDTO;
import com.joelcode.personalinvestmentportfoliotracker.entities.Order;

import java.util.List;
import java.util.UUID;

public interface OrderService {

    /**
     * Create a new limit order
     */
    OrderDTO createOrder(CreateOrderRequest request);

    /**
     * Get a specific order by ID
     */
    OrderDTO getOrderById(UUID orderId);

    /**
     * Get all orders for an account (all statuses)
     */
    List<OrderDTO> getAllOrdersForAccount(UUID accountId);

    /**
     * Get all pending orders for an account
     */
    List<OrderDTO> getPendingOrdersForAccount(UUID accountId);

    /**
     * Cancel a pending order
     */
    OrderDTO cancelOrder(UUID orderId);

    /**
     * Get all pending orders across all accounts (for scheduler)
     */
    List<Order> getAllPendingOrders();

    /**
     * Mark an order as executed after transaction is created
     */
    void markOrderAsExecuted(UUID orderId, UUID transactionId);

    /**
     * Mark an order as failed due to insufficient balance/shares
     */
    void markOrderAsFailed(UUID orderId, String reason);
}

File: frontend/src/services/orderService.ts
import { apiFetch } from './api';
import type { OrderDTO, CreateOrderRequest } from '../types/api';

/**
 * Create a new limit order
 * POST /api/orders
 */
export const createOrder = async (order: CreateOrderRequest): Promise<OrderDTO> => {
  return apiFetch<OrderDTO>('/api/orders', {
    method: 'POST',
    requireAuth: true,
    body: JSON.stringify(order),
  });
};

/**
 * Get a specific order by ID
 * GET /api/orders/{orderId}
 */
export const getOrderById = async (orderId: string): Promise<OrderDTO> => {
  return apiFetch<OrderDTO>(`/api/orders/${orderId}`, {
    method: 'GET',
    requireAuth: true,
  });
};

/**
 * Get all orders for an account (all statuses)
 * GET /api/orders/account/{accountId}
 */
export const getAccountOrders = async (accountId: string): Promise<OrderDTO[]> => {
  return apiFetch<OrderDTO[]>(`/api/orders/account/${accountId}`, {
    method: 'GET',
    requireAuth: true,
  });
};

/**
 * Get all pending orders for an account
 * GET /api/orders/account/{accountId}/pending
 */
export const getPendingOrders = async (accountId: string): Promise<OrderDTO[]> => {
  return apiFetch<OrderDTO[]>(`/api/orders/account/${accountId}/pending`, {
    method: 'GET',
    requireAuth: true,
  });
};

/**
 * Cancel a pending order
 * DELETE /api/orders/{orderId}
 */
export const cancelOrder = async (orderId: string): Promise<OrderDTO> => {
  return apiFetch<OrderDTO>(`/api/orders/${orderId}`, {
    method: 'DELETE',
    requireAuth: true,
  });
};
